src/gedi_geoparquet/hdf5.py

```diff
@@ -0,0 +1,344 @@
+from __future__ import annotations
+
+import typing as t
+from enum import StrEnum, auto
+
+import h5py
+import numpy as np
+import polars as pl
+import pyarrow as pa
+
+import gedi_geoparquet.polars_ as pl_
+
+
+class BeamType(StrEnum):
+    """GEDI beam types.
+
+    Within each HDF5 file, regardless of GEDI collection, there are eight top
+    level groups (see ``BeamName`` for their names), each falling into one of
+    two beam types, which are named in the "description" attribute of each beam
+    group as either "Coverage beam" or "Full power beam".
+
+    Here we enumerate them more succinctly as either "coverage" or "power",
+    following what users commonly term them.
+    """
+
+    COVERAGE = auto()
+    POWER = auto()
+
+
+class BeamName(StrEnum):
+    """GEDI Beam names.
+
+    Across all GEDI collections, there are eight commonly named beams, appearing
+    as top level groups within GEDI HDF5 files.
+
+    They are always associated with specific beam types (see ``BeamType``).
+    """
+
+    type: BeamType
+
+    BEAM0000 = auto(), BeamType.COVERAGE
+    BEAM0001 = auto(), BeamType.COVERAGE
+    BEAM0010 = auto(), BeamType.COVERAGE
+    BEAM0011 = auto(), BeamType.COVERAGE
+    BEAM0101 = auto(), BeamType.POWER
+    BEAM0110 = auto(), BeamType.POWER
+    BEAM1000 = auto(), BeamType.POWER
+    BEAM1011 = auto(), BeamType.POWER
+
+    def __new__(cls, value: str, type_: BeamType) -> t.Self:
+        member = str.__new__(cls, value)
+        # For StrEnum types, `auto()` sets values to the lowercase of the keys,
+        # so we're overriding that here to ensure uppercase keys, because the
+        # group names in the HDF5 files are uppercase.
+        member._value_ = value.upper()
+        member.type = type_
+        return member
+
+
+EPOCH_NS = np.datetime64("2018-01-01T00:00:00", "ns")
+"""Epoch for the GEDI mission.
+
+All `delta_time` datasets are time deltas (in seconds) since this epoch.  When
+converting `delta_time` float values to timedelta values to add to this epoch to
+compute absolute times, multiply by 1e9 (to convert from float seconds to
+nanoseconds), convert to `int`, and use `ns` (nanosecond) resolution.
+"""
+
+
+def to_arrow(file: h5py.File, *, schema: pa.Schema) -> pa.Table:
+    """Read a GEDI HDF5 file into an Arrow Table.
+
+    This is the Arrow equivalent of the materialized Polars LazyFrame produced
+    by the ``to_polars`` function.  See that function for details.
+    """
+
+    schema_metadata = schema.metadata if schema else None
+    polars_schema = pl.Schema(schema)
+
+    # WARNING: When Polars converts to Arrow, it defaults to using LargeString and
+    # LargeList even when String and List will suffice.  This might be problematic.
+    # See https://github.com/pola-rs/polars/issues/15047.
```

Comment:
This isn't true anymore I'm pretty sure. Polars uses the new "string view" type for all its data, which other Arrow implementations also support.